File: wordSearch/wordSearch.cpp
#include <iostream>
#include <vector>
#include <list>
using namespace std;

class Solution{
public:
  bool exist(vector<vector<char> >& board, string word) {
    for (int y = 0; y < board.size(); y++) {
      for (int x = 0; x < board[y].size(); x++) {
	if (findFromHere(board, x, y, word, 0))
	  return true;
      }
    }
    return false;
  }
  bool findFromHere(vector<vector<char> >& board, int x, int y, string word, int depth) {
    if (depth == word.length())
      return true;
    if (y < 0 || x < 0 || y == board.size() || x == board[y].size())
      return false;
    if (board[y][x] != word[depth])
      return false;
    board[y][x] = '*'; 
    bool find = findFromHere(board, x+1, y, word, depth+1)
	    || findFromHere(board, x-1, y, word, depth+1)
	    || findFromHere(board, x, y+1, word, depth+1)
	    || findFromHere(board, x, y-1, word, depth+1);
    board[y][x] = word[depth];
    return find;
  }
};

int main()
{
  Solution s;
  char a0[4] = {'A', 'B', 'C', 'E'};
  char a1[4] = {'S', 'F', 'C', 'S'};
  char a2[4] = {'A', 'D', 'E', 'E'};
  vector<char> vec0(a0, a0+4);
  vector<char> vec1(a1, a1+4);
  vector<char> vec2(a2, a2+4);  
  vector<vector<char> > vec;
  vec.push_back(vec0);
  vec.push_back(vec1);
  vec.push_back(vec2);
  string word = "ABCCED";
  cout << s.exist(vec, word);
  cout << endl;
  return 0;
}

File: wordSearch/wordSearch.java
class Solution {
    public boolean exist(char[][] board, String word) {
        int m = board.length, n = board[0].length;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (board[i][j] == word.charAt(0)) {
                    if (dfs(board, word, 0, m, n, i, j)) return true;
                }
            }
        }
        return false;
    }
    
    private boolean dfs(char[][] board, String word, int level, int m, int n, int i, int j) {
        if (level == word.length()) {
            return true;
        }
        char c = word.charAt(level);
        if (i < 0 || i >= m || j < 0 || j >= n || c != board[i][j]) {
            return false;
        }
        board[i][j] = '#';
        boolean res = dfs(board, word, level + 1, m, n, i + 1, j)
            || dfs(board, word, level + 1, m, n, i - 1, j)
            || dfs(board, word, level + 1, m, n, i, j + 1)
            || dfs(board, word, level + 1, m, n, i, j - 1);
        board[i][j] = c;
        return res;
    }
}
